hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/impl/listener/ClientNonSmartListenerService.java

```diff
@@ -23,100 +23,134 @@
+
+        Address address = future.getInvocation().getSendConnection().getRemoteEndpoint();
+        return new ClientEventRegistration(registrationId,
+                addRequest.getCallId(), address, removeRequest);
+
     }
 
+
     @Override
-    public boolean deregisterListener(BaseClientRemoveListenerRequest request, String userRegistrationId) {
-        ClientEventRegistration registration = registrations.remove(new ClientRegistrationKey(userRegistrationId));
-        if (registration == null) {
-            return false;
-        }
-        removeEventHandler(registration.getCallId());
-        request.setRegistrationId(registration.getServerRegistrationId());
-        try {
-            Future future = new ClientInvocation(client, request, registration.getSubscriber()).invoke();
-            future.get();
-        } catch (Exception e) {
-            //if invocation cannot be done that means connection is broken and listener is already removed
-            EmptyStatement.ignore(e);
+    public boolean deregisterListener(String userRegistrationId) {
+        synchronized (listenerRegLock) {
+            ClientRegistrationKey key = new ClientRegistrationKey(userRegistrationId);
+            ClientEventRegistration registration = registrations.get(key);
+
+            if (registration == null) {
+                return false;
+            }
+
+            BaseClientRemoveListenerRequest removeRequest = registration.getRemoveRequest();
+            removeRequest.setRegistrationId(registration.getServerRegistrationId());
+            try {
+                Future future = new ClientInvocation(client, removeRequest).invoke();
+                future.get();
+                removeEventHandler(registration.getCallId());
+                registrations.remove(key);
+            } catch (Exception e) {
+                throw new HazelcastException("Listener with id " + userRegistrationId + " could not be removed", e);
+            }
+            return true;
         }
-        return true;
     }
 
     @Override
-    public void memberAdded(MembershipEvent membershipEvent) {
-        Address ownerConnectionAddress = client.getClientClusterService().getOwnerConnectionAddress();
-        if (membershipEvent.getMember().getAddress().equals(ownerConnectionAddress)) {
-            executionService.executeInternal(new Runnable() {
-                @Override
-                public void run() {
+    public void connectionAdded(Connection connection) {
+        executionService.executeInternal(new Runnable() {
+            @Override
+            public void run() {
+                synchronized (listenerRegLock) {
                     for (ClientRegistrationKey registrationKey : registrations.keySet()) {
-                        invoke(registrationKey);
+                        try {
+                            ClientEventRegistration registration = invoke(registrationKey);
+                            registrations.put(registrationKey, registration);
```

Comment:
Lets assume, I removed the lock from the invocations, 
Lets say we have connection added and deregister together. Deregister get the lock removed it from map still holding the lock. 

Then connection added starting sending invocations to remote which will be completely wrong registrations.  Locks purpose is violated here.